### DISCUSSION TITLE
How to avoid partially unknown return type for generic list?

### DISCUSSION THREAD

**Discussion by jawshooah** [microsoft/pyright#4148] [Q&A] [answered]
How would I modify this code to avoid `error: Return type, "list[Unknown]* | list[T@ensure_list]", is partially unknown (reportUnknownVariableType)`?

```python
# pyright: strict
from typing import TypeVar

T = TypeVar("T")


def ensure_list(value: T | list[T] | None) -> list[T]:
    if value is None:
        return []
    if not isinstance(value, list):
        return [value]
    return value
```

Here's what `pyright` gives me:
```
❯ pyright type_partially_unknown.py
No configuration file found.
No pyproject.toml file found.
stubPath /Users/jhagins/typings is not a valid directory.
Assuming Python platform Darwin
Searching for source files
Found 1 source file
pyright 1.1.278
/Users/jhagins/type_partially_unknown.py
  /Users/jhagins/type_partially_unknown.py:12:12 - error: Return type, "list[Unknown]* | list[T@ensure_list]", is partially unknown (reportUnknownVariableType)
1 error, 0 warnings, 0 informations
Completed in 0.545sec
```

**Comment by erictraut** [comment, marked as answer]
I don't see a good way to avoid this. Any time you create a union with a "naked" TypeVar and other types that use that TypeVar as a type argument, you run the risk of hitting situations like this because the rules for constraint solving (i.e. determining the value of the TypeVar) and type narrowing are ambiguous. Pyright has a bunch of heuristics in both of these code paths that attempt to come up with the "best" solution in the majority of cases, but there will always be edge cases that don't work as well.

In this particular case, the type narrowing for `isinstance(value, list)` when `value` has the (already narrowed) type of `T | list[T]` produces the type `list[Unknown]* | list[T]` in the positive ("if") case and `object*` in the negative ("else") case. (The `not` operator effectively flips the positive and negative results.)

You could use `return cast(list[T], value)` or use `# pyright: ignore[reportUnknownVariableType]`.
